janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java

```diff
@@ -82,55 +96,31 @@ public void setUseMultiQuery(boolean useMultiQuery) {
         return (Iterator<E>) Iterators.transform(iterable.iterator(), Property::value);
     }
 
-    private void initialize() {
-        assert !initialized;
-        initialized = true;
-        assert getReturnType().forProperties() || (orders.isEmpty() && hasContainers.isEmpty());
-
-        if (!starts.hasNext()) throw FastNoSuchElementException.instance();
-        final List<Traverser.Admin<Element>> elements = new ArrayList<>();
-        starts.forEachRemaining(elements::add);
-        starts.add(elements.iterator());
-        assert elements.size() > 0;
-
-        useMultiQuery = useMultiQuery && elements.stream().allMatch(e -> e.get() instanceof Vertex);
-
-        if (useMultiQuery) {
-            initializeMultiQuery(elements);
-        }
-    }
-
     /**
-     * This initialisation method is called the first time this instance is used and also when
-     * an attempt to retrieve a vertex from the cached multiQuery results doesn't find an entry.
-     * @param vertices A list of vertices with which to initialise the multiQuery
+     * This initialisation method is called when an attempt to retrieve a vertex from the cached multiQuery results
+     * doesn't find an entry.
      */
-    private void initializeMultiQuery(final List<Traverser.Admin<Element>> vertices) {
-        assert vertices.size() > 0;
-        final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
-        vertices.forEach(v -> multiQuery.addVertex((Vertex)v.get()));
+    private void prefetchNextBatch() {
+        final JanusGraphMultiVertexQuery multiQuery = makeMultiQuery(batchSuppliers);
         makeQuery(multiQuery);
 
-        Map<JanusGraphVertex, Iterable<? extends JanusGraphProperty>> results = multiQuery.properties();
-        if (multiQueryResults == null) {
-            multiQueryResults = results;
-        } else {
-            multiQueryResults.putAll(results);
+        try{
+            Map<JanusGraphVertex, Iterable<? extends JanusGraphProperty>> results = multiQuery.properties();
+            if (multiQueryResults == null) {
+                multiQueryResults = results;
+            } else {
+                multiQueryResults.putAll(results);
```

Comment:
Can we re-initialize multiQueryResult instead of putAll? What is case when some elements will be retrieved repeatedly?
If we can re-initialize it significantly reduce memory usage for OLAP query. In some cases more preferable execute another query for repeated elements instead of OOM

Reply:
Very interesting thought! Just from the top of my head, I imagine this could cause problems with MultiQuerieables within `RepeatStep`, where the `multiQueryResults` are not necessarily used in the same order in which they are retrieved. But I will of course look into that!

Reply:
Turns out I was probably wrong about the execution order of the repeated traversals. Could possibly work even better than I expected.